system/security/shared/caching.cpp

```diff
@@ -672,16 +680,55 @@ bool CPermissionsCache::queryPermsManagedFileScope(ISecUser& sec_user, const cha
 
 SecAccessFlags CPermissionsCache::queryDefaultPermission(ISecUser& user)
 {
-    if (m_defaultPermission == SecAccess_Unknown)
+
+    if (m_useLegacyDefaultFileScopePermissionCache)
     {
-        if (m_secMgr)
-            m_defaultPermission = m_secMgr->queryDefaultPermission(user);
+        if (m_defaultPermission == SecAccess_Unknown)
+        {
+            if (m_secMgr)
+                m_defaultPermission = m_secMgr->queryDefaultPermission(user);
+            else
+                m_defaultPermission = SecAccess_Full;
+
+            DBGLOG("Legacy default file scope permission set to %s(%d) for all users, based on User '%s'", getSecAccessFlagName(m_defaultPermission),
+                   m_defaultPermission, user.getName());
+        }
+        return m_defaultPermission;
+    }
+
+    SecAccessFlags defaultPermission = SecAccess_None;
+    if (m_secMgr)
+    {
+        const std::string username(user.getName());
+        ReadLockBlock defaultScopesReadLock(m_defaultScopesRWLock);
```

Comment:
I would have used a simpler critical section.
- The R/W lock complicates this code (relock and potential for deadlock you have avoided).
- This is unlikely to be heavily contended.
- R/W lock is less efficient than a a critical section.
- If it contended and needs a write then the other threads will need to wait anyway.
The R/W blocks are good for situations where parallel reads can continue and not have the write block them.  (Even then it is only likely to be helpful when slow and heavily contended.)

Reply:
From what I have seen, it appears that default file scopes are used a lot. Making this code as efficient as possible is probably the right thing to do. I agree that lock for read, release, lock for write and check again, is not optimal. Since you mentioned a critical section, I think the time to change it is now. I will make that change in the interest of keeping default scope checking efficient.